### ISSUE TITLE
Missing from VLE is Forum Contribution

### ISSUE THREAD

**Issue by arc12** [jiscdev/xapi#132]
We can see resource viewing which happens to be a forum, but we have no idea about students (or others) posting/commenting.

This rather constrains our view on active participation.

I accept that good use of forums is the exception rather than the rule, but still...

**Comment by ds10** [comment]
@arc12 I've spent some time thinking about this and added some things for us to talk around in to the 1.0 branch. I guess there are different approachs.

- In the two files I have added there are two statements, with verbs Posted and Replied. Each of these has an object which is the post, a result with the posted text, context with courseArea, session id etc. Do 

- I'm not sure about how far the vle breaks things down. Should the object in the posted statement be the post itself and the object in the replied be the comment? If so then the statements need a way of refering to each other. I guess the context.contextactivities.parent would refer to the forum the post was made in or the post the comment was made in? So, the reply to the post and the post's parent is the forum?

- I guess another way would be to have reply statements as statementreferences to the original post statement?

I have just put two examples in 1.0 to get us going, @arc do you want to comment?

**Comment by arc12** [comment]
Hmm... forum events/lifecycle tends to end up being non-trivial...

tl;dr: lets not worry about the relational structure and theoretical semantics.

My inclination is to use only a "posted" verb and to leave it to anyone who might process the data to work out whether it was really a reply to another post, a contribution within the correct scope of a thread, or a spam/non-sequitur post. I remember from our work with the coursera data that the semantics of post/reply/comment was often mangled in the real data. What was really a reply didn't necessarily show up as a reply in the DB and what was supposed to be a reply according to the DB was clearly not. It is clear to me that the UI of the forum is influential here, but we'll generally either be ignorant of this or unprepared to write differentiating code. Hence, we'll probably end up processing posts as a set of posts in a bucket. Frankly, in the first instance, we'll just be counting posts under modules.

I agree that the posted text should appear in the statement. Not sure where. As the description of the object.definition? We won't be doing text analytics in the short term but it would be silly to preclude it.

I have seen people talking about using result for the responses of other people, but I don't really like this. Result doesnt feel right for a forum post.

Agree with using context. I would allow for forum and thread identifiers and we need to define the concepts of "starting a thread" vs "contributing to a thread" in reltion to the absence/presence of the relevant identifier. Probably will need to explain terms too, with some synonyms like "board", "topic". I'd be inclined to double-code these relational identifiers: to use both contextActivities parent and grouping and some jisc.ac.uk extensions

Statementref seems like a sensible way of referring to the post which was replied to but:-
same challenge as you raised re statement voiding: the source system may not know the LRS id and a query is probably impractical. So... maybe an in-reply-to context extension which has the object id as known to the source vle for the message which it thinks this is a reply to.

Someone really should have solved this problem!
(but IMO the SemWeb models I've seen are unrealistically precise)

**Comment by alanepaull** [comment]
I suspect that decisions about these details depend very much on what the purpose of the data record is. Are we recording and analysing 'participation'? So, this student has posted loads of stuff in this forum, but this other student hasn't? Or do we care (a lot) about *what* the student has posted?

If we're not going to analyse the text or the position within a thread, we don't need (necessarily) to hold it. Perhaps we just need the metadata, not the content?

I agree that excessive precision may be excessive :) for our purposes.

**Comment by arc12** [comment]
my intuitive hierarchy of need is:

1. THAT the student posted
2. WHERE the student posted (broad contextual)
3. WHAT the student posted (or staff, lets not forget*)
4. in-reply-to and detailed relational

I think losing the text would be a missed future opportunity. Lots of scope for simple text mining: word count, keyword use, sentiment, ...

* - for this and other cases, we probably should know the actor role from the statement. In principle, we should be able to find this out if we have UDD data for staff and students which gives us a staff and student identifier lookup BUT BUT BUT this is quite unclean and often problematical data to access. A simple roletype indicator would help to reduce complexity. It would also help to guarantee we dont inadvertantly include staff in summative statistics for module activity in those cases where pre-joining to UDD is not going on

**Comment by ds10** [comment]
I like the hierarchy...

I had quite a look at this yesterday and your fourth point was the hardest. Mostly because there are so many ways to tackle it, which really explains the approach I took.

It would be much easier to have every statement using the post verb in the forum, rather than post in a forum + comment on a post

**Comment by ds10** [comment]
OK @arc12 . I've put two templates up for discussion. Posted and replied. Both are the same except the verb. The object is the post, but the umberella forum can be found in the context.

Do you want to have a look at these two and we can take it from there? Can be found in v 1.0 branch under VLE recipe

**Comment by ds10** [comment]
From Monday Call:
Adam to check

**Comment by ds10** [comment]
@arc12 are you still ok to have a look at these forum contribution templates? They can be found in 1.0 branch

**Comment by arc12** [comment]
OK... feedback part 1 - simple:

Clarify that result.response is the plain text content of the post, properly encoded as per the encoding of the rest of the statement (UTF-8 I hope!). NO markup. NO HTML entity encodings. NO invalid utf8.

Could both templates be collapsed to one which just says that the verb differs according to whether the post starts or continues a discussion?

**Comment by arc12** [comment]
part 2 - not so simple:

I think we need to be more clear about the object model, both in outline and in relation to the component parts of the statement in the documentation.

Current thoughts on this:
1. clarify a hierarchy comprising forum > discussion (aka thread) > post.
2. "posted" verb starts a discussion, "replied" continues it.
3. identify the discussion/thread as the object for all statements. It is essential for sanity in reconstructing patterns for this not to be a mix of different entities according to the guesswork of plugin developers.
4. identify the parent forum as context. I would like this to be accessible without dealing with contextActivities, i.e. with a jisc-minted URI extension
5. I'm not that bothered by knowing the "in-reply-to-post-id", but some customers may want to be able to attempt to reconstuct sequence.
6. key things to be able to query or group data on - agent-id, vle-mod-id, forum, discussion/thread - need to be in nice consistent locations. First two already are.

**Comment by alanepaull** [comment]
For @arc12, just for information... is contextActivities.parent problematic? That's the "normal" (?) way that this relationship might be done in xAPI, I think.

**Comment by arc12** [comment]
@alanepaull one practical problem is that the xAPI spec defines this as an array, which introduces some doubt into extracting "what is THE parent forum".

**Comment by ds10** [comment]
@arc Thanks for this, I will have a look asap, but those comments make sense. I think we decided monday that contextActivities was out anyway.

**Comment by alanepaull** [comment]
@arc12 Indeed it does, even if it could be an array of 1 :).

**Comment by ds10** [comment]
Ok, here is my attempt from your comments: https://github.com/jiscdev/xapi/blob/1.0/recipes/vle/forum.md

There is not description yet and common statements and vocab have not been updated yet:

1. This is now just one template, with the option of verb for posted or replied
2. The object is the forum post
3. tentatively put a forumArea property in the context
4. Put a note on result about markup, this would be mirrored in the common_statements
5. Removed contextactivties

**Comment by arc12** [comment]
documentation for context in forum.md needs work.

refers to "contextX" which does not exist
query ""http://jisc.ac.uk/forumArea" - should be an extension, probably with xapi.jisc.ac.uk URI

**Comment by alanepaull** [comment]
The tincan documentation encourages us not to mint new verbs if possible. Therefore, I'm wondering whether we should use an existing verb, rather than mint "posted".

Existing alternatives might be:

commented URI: http://adlnet.gov/expapi/verbs/commented - Offered an opinion or written experience of the activity. Can be used with the learner as the actor or a system as an actor. Comments can be sent from either party with the idea that the other will read and react to the content.

created URI: http://activitystrea.ms/schema/1.0/create - Indicates that the actor has created the object.

added URI: https://brindlewaye.com/xAPITerms/verbs/added/ - Added one or more items to a collection

**Comment by alanepaull** [comment]
Tincan already has these covered, I feel, with these Activity types:

Forum Topic URI: http://id.tincanapi.com/activitytype/forum-topic - The first post in a thread on a forum or discussion board.

Forum Reply URI: http://id.tincanapi.com/activitytype/forum-reply - Any post in a forum or discussion board thread that isn't the first.

There is also:
discussion URI: http://id.tincanapi.com/activitytype/discussion - Represents an ongoing conversation between persons, such as an email thread or a forum topic.

David and Adam: shall we look at incorporating these into forum.md?

**Comment by alanepaull** [comment]
I've drafted a version of what I mean in forum_alternative.md, now posted to: 

https://github.com/jiscdev/xapi/blob/1.0/recipes/vle/forum_alternative.md.

Feel free to discuss/amend/reject!

**Comment by arc12** [comment]
I'm not too particular about the choice of verb. there is a balance between being abstract and widely used vs being more contextually-refined. You create many things vs posting to a forum. Over-abstraction makes for lack of clarity when data processing. over-refinement leads to tautological verb+objectType. I could live with "created" and a clearly forum-specific object.definition.type.

However, the alt template has both verb and object.definition.type doing the same duty of disciminating between initial post and reply. I think it would be cleaner to use  http://id.tincanapi.com/activitytype/discussion here.

Colloquially:
- a person creates a discussion in the context of forum F
- a person replies to the discussion in ...

If we want to query for forum contribution, we just look for http://id.tincanapi.com/activitytype/discussion

HOWEVER!
we seem to have forgotten about the use case "query for forum contribution and consumption in one simple way"
This drives us towards making http://xapi.jisc.ac.uk/vle/forum as the object.definition.type and moving the identifier for the thread/discussion into context (as optional, to support cases where there is no threading).
In that case, we would NOT want to use "created" but to be specific about posting.

Also, I think we should note that software which cannot identify a contribution as being a reply in a thread/discussion should always use posted, and should omit the discussion/thread identifier (assumed to be) in context.

So... still some open questions!

**Comment by alanepaull** [comment]
We need to identify the forum, the discussion thread and the posts, possibly only identifying the initial post and not identifying the replies individually, but knowing that they are replies in that discussion thread.

We need an easy way to query forum contributions.

Natural language example: John Smith goes to a forum in the VLE. He wants to start a new discussion in that forum, so he writes a new piece of text and posts it, creating a new discussion thread. Jenny Jones writes a comment in response to John Smith's piece and posts her reply. Several other students add their replies, both to the original piece and to subsequent replies.
In summary: John Smith starts a new discussion thread by posting in a forum. Jenny Jones (and others) post replies to this discussion thread in the forum.

It's difficult to have the discussion or the forum as the object of a "John Smith posts..." style statement, because the verb is "doing something" to/with "something". The natural "something" is the text that the Actor is writing. The discussion thread is a collection of comments; a forum is a collection of discussion threads. 
Forum
     Discussion thread
          Initial post }  comment or post
          Reply post }

If we want to retain this structure, then I suggest that we need to make the comment or post the object of a suitable verb, "the act of posting/replying/commenting". Then the discussion thread and forum sit in the context.

Fortunately, this has been done before: http://xapiquarterly.com/2016/10/instrumenting-xapi-in-forums-discussion-groups/

For initial posts: 
![image](https://user-images.githubusercontent.com/17160871/27920190-bd7d6742-626b-11e7-882d-ad7ed76eeeb4.png)

And for replies: .
![image](https://user-images.githubusercontent.com/17160871/27920104-78bcdf02-626b-11e7-9b5e-d38bcbb57f48.png)

I'm not quite sure if we can learn from this, but I thought it was worth a go.  Not yet had time to digest it myself.

**Comment by alanepaull** [comment]
In these examples they've used the verb to differentiate between the initial post ("posted") and other comments ("replied"). The object.definition.type in each case is "forum-topic". The forum is identified in Context as contextActivities.parent using the tincan "discussion" type. They've also specified the replies as part of a contextActivities.grouping with type "forum-reply". These all look to me like very standard xAPI structures.

The actual post content hasn't been captured, but I guess that *could* be captured in result.response, as we've suggested elsewhere.

These examples use the contextActivities structures, which may not be palatable. However, it does the job. Do we have an alternative, better solution? Do we need to seek one?

**Comment by ds10** [comment]
I think this is similar to the approach I had taken in my first two examples. I had two statements using the verb to differentiate - posted and replied and used contextActivities grouping in a similar fashion - I had  put the content in result.response. 

I think the response was to avoid contextActivities structures where possible, so in my second attempt (that was in a single template) I kept the two verbs but removed contextActivies. We had a discussion about using http://xapi.jisc.ac.uk/vle/forum (which doesn't seem right) as the object.type and minting http://xapi.jisc.ac.uk/forumArea to identify the parent forum.

**Comment by alanepaull** [comment]
Yes, very similar. I thought I'd post those examples, because they are directly an approach that's already in use. I'm very loathe to use a different approach to one already in use, if we can avoid it. The main difference between these examples and our / your earlier one is that they've used "standard" bits like forum-topic and discussion that I like.

Can we confirm whether or not we definitely shouldn't be using contextActivities? If we know that, we can go from there. Using parent makes identifying the wider context easy (in theory) - but we don't do that with courseArea, so maybe stylistically and for practical reasons (use of arrays), we don't want to do it.

**Comment by alanepaull** [comment]
Monday 17 July: Agreed to use contextActivities.parent, generally, BUT that Jisc Profile should specify 1 entry only, in other words not an array or more than 1.

We didn't confirm that we would use the standard tincanapi terms. I propose to use these terms:

- Forum = http://id.tincapi.com/activitytype/discussion, used within contextActivities.parent where appropriate
- Forum thread = http://id.tincapi.com/activitytype/forum-topic
- Verb for posting an initial post (new thread) = https://w3id.org/xapi/acrossx/verbs/posted
- Verb for posting a reply on a forum = http://id.tincanapi.com/verb/replied

I'll modify forum-alternative.md to cover this.

Alan

**Comment by arc12** [comment]
I feel there are still some structural questions which seem to be getting away from us, and with two candiates in markdown + one from the web + lots of issue comments, its not surprising!

I think this needs a reboot, with all the material so far created "to hand" as raw material.

I also think we need to clarify principles before-hand. I have two-plus-one:
1. compatible structure to vle forum view statement so passive and active forum access can be trivially queried for together
2. coherence with xAPI specification and sensible re-use of existing URIs
3. avoidance of assumption that theoretical-semantic distinctions apply in practice (e.g. that it is possible to know which post a reply was in relation to)
 
As concrete concerns go, I feel we've lost sight of # 1

**Comment by alanepaull** [comment]
@arc12: It would be helpful if you would post a vle forum view statement, I think, should you have a sample to hand. We have these view examples already, but they're not forums:

https://github.com/jiscdev/xapi/blob/1.0/vle/moodle/moduleview.js
https://github.com/jiscdev/xapi/blob/1.0/vle/blackboard/course_content_access.json

I don't know if the module_view (https://github.com/jiscdev/xapi/blob/1.0/recipes/vle/Module-View.md) covers everything we need for forum view. I don't think it does, because it's more specific to module information.

I agree that we should clarify principles; that seems like a sound basis for discussion and eventual agreement.

I think your #1 may be tricky, but I don't want to get into a discussion about what "compatible" might mean! I think that we're seeking to get a readily identifiable structural component to identify the forum for querying.

If we record "viewing a forum", do we mean simply that the student has viewed any part of the forum? Just the title page? Or a specific discussion thread or even a specific post or reply, or sequence of replies? Any of these might be represented simply by a generic forum object, if we don't care which bit they viewed, but just that they viewed the forum area. If we do care, we can still have some mechanism to identify the forum in context.

Taking requirement #1, perhaps all our forum-related statement templates should have the forum as the contextActivities.parent. So any query about a specific forum can pick them up easily through that structure.

#2 may be covered by previous comments about using existing vocabularies from elsewhere (like forum-topic) that seem to have some provenance, as long as they work the way we need.

#3 might be a plea not to try to model forum structures. I would agree.

I think that #1 may be the most important issue here, and if we can agree that, then the rest may drop out with good will, flexibility and agreement that we don't want to run before we can walk. I'm happy to draft a couple of bits (forum view and forum posting) on the basis of previous work and this discussion. Probably just snippets in the next couple of days to see if we get the basis of agreement, rather than whole templates.

**Comment by arc12** [comment]
Example from the LRW (with some ID changed)

>             "statement": {
>                 "version": "1.0.1",
>                 "timestamp": "2013-06-28T13:05:12+01:00",
>                 "object": {
>                     "objectType": "Activity",
>                     "definition": {
>                         "extensions": {
>                             "http://xapi.jisc.ac.uk/applicationType": "http://id.tincanapi.com/activitytype/lms"
>                         },
>                         "type": "http://xapi.jisc.ac.uk/vle/forum",
>                         "description": {
>                             "en": "A module"
>                         },
>                         "name": {
>                             "en": "HSOC"
>                         }
>                     },
>                     "id": "http://www.dglearnnet.com/moodle3/forum/view.php?id=289"
>                 },
>                 "actor": {
>                     "objectType": "Agent",
>                     "account": {
>                         "homePage": "http://www.dglearnnet.com/moodle3/",
>                         "name": "Adam"
>                     }
>                 },
>                 "verb": {
>                     "id": "http://id.tincanapi.com/verb/viewed",
>                     "display": {
>                         "en": "viewed"
>                     }
>                 },
>                 "context": {
>                     "platform": "Moodle",
>                     "extensions": {
>                         "http://xapi.jisc.ac.uk/recipeVersion": "1.0",
>                         "http://xapi.jisc.ac.uk/sessionId": "n/a",
>                         "http://lrs.learninglocker.net/define/extensions/info": "Moodle",
>                         "http://xapi.jisc.ac.uk/courseArea": {
>                             "id": "http://www.dglearnnet.com/moodle3/forum/view.php?id=289",
>                             "http://xapi.jisc.ac.uk/vle_mod_id": "HSOC-ONLINE"
>                         }
>                     }
>                 },
>                 "authority": {
>                     "objectType": "Agent",
>                     "name": "New Client",
>                     "mbox": "mailto:hello@learninglocker.net"
>                 },
>                 "stored": "2017-05-18T14:13:47.602100+00:00",
>                 "id": "7a88bae6-9116-44b7-ae69-5811e5157ba1"
>             }

**Comment by arc12** [comment]
the current statement for vle "module view" (I think this name is slightly misleading and arose from the Moodle use of "module") is, indeed, very blunt, with no detail other than accessing the forum somehow. Depending on whether the VLE and plugin it may or may not be possible to get more detail, and for non-online courses, I suspect we would gain little insight; in terms of mainstream HE/FE data at scale, conversation maps and SNA are a pure pipe-dream.

**Comment by alanepaull** [comment]
If we accept that we need a consistent way to indicate a statement that relates to a forum in a VLE, I suggest that we consider this type of structure in Context:

```
"context": {
  "platform": "Moodle",
  "extensions": {
    "http://xapi.jisc.ac.uk/sessionId": "123456",
    "http://xapi.jisc.ac.uk/courseArea": {
      "http://xapi.jisc.ac.uk/vle_mod_id": "HSOC-ONLINE", 
      "http://xapi.jisc.ac.uk/uddModInstanceID": "HSOC01-2016S1-0"
    }
  }
  "contextActivities": {
    "parent": [
      {
        "objectType": "Activity",
        "id": "http://www.dglearnnet.com/moodle3/forum/view.php?id=289",
        "definition": {
          "name": {
            "en": "HSOC forum"
          },
          "type": "http://id.tincanapi.com/activitytype/discussion"
        }
      }
    ]
  }
}
```
It may be that the courseArea extension isn't necessary, so should be optional.

**Comment by alanepaull** [comment]
For a forum initial post, we could use "create" as the verb and "forum-topic" as the object:
```
"verb": {
        "id": "http://activitystrea.ms/schema/1.0/create",
        "display": {
            "en": "created"
        }
    }

"object": {
   "objectType": "Activity",
   "id": "http://moodle2.bolton.ac.uk/mod/forum/discuss.php?d=19474"	
   "definition": {
   	"type": "http://id.tincanapi.com/activitytype/forum-topic",			
   	"name": { "en": "Forum topic" },			   
    }
   }
}
```

**Comment by alanepaull** [comment]
For a forum reply, we have a choice. We could use the same verb and change the object to "forum-reply":
```
"verb": {
        "id": "http://activitystrea.ms/schema/1.0/create",
        "display": {
            "en": "created"
        }
    }

"object": {
   "objectType": "Activity",
   "id": "http://moodle2.bolton.ac.uk/mod/forum/discuss.php?d=19474"	
   "definition": {
   	"type": "http://id.tincanapi.com/activitytype/forum-reply",			
   	"name": { "en": "Forum reply" },			   
    }
   }
}
```
Or we could change the verb to "replied" and use the same "forum-object" object.
```
"verb": {
        "id": "http://id.tincanapi.com/verb/replied",
        "display": {
            "en": "replied"
        }
    }

"object": {
   "objectType": "Activity",
   "id": "http://moodle2.bolton.ac.uk/mod/forum/discuss.php?d=19474"	
   "definition": {
   	"type": "http://id.tincanapi.com/activitytype/forum-topic",			
   	"name": { "en": "Forum topic" },			   
    }
   }
}
```
Or we could have some redundancy and change both verb and object:
```
"verb": {
        "id": "http://id.tincanapi.com/verb/replied",
        "display": {
            "en": "replied"
        }
    }

"object": {
   "objectType": "Activity",
   "id": "http://moodle2.bolton.ac.uk/mod/forum/discuss.php?d=19474"	
   "definition": {
   	"type": "http://id.tincanapi.com/activitytype/forum-reply",			
   	"name": { "en": "Forum reply" },			   
    }
   }
}
```
All of these snippets would have the Context as above, so that the forum can be queried readily (with the optional courseArea).

Personally, I think I favour the reply option with verb=replied and object=forum-topic. Then someone in future could theoretically pull out the discussion thread. It allows identification of initial posts through "create", replies through "replied", threads through "forum-topic", forums through contextActivities.parent, and other stuff through courseArea.

**Comment by arc12** [comment]
The problem I have with the object.definition.type being a forum topic is that it is not the same as the existing vle/forum view has. I feel it would be very helpful to have a consistent object type for the variety of activity-in-the-forum statements.

**Comment by alanepaull** [comment]
I accept that my proposal is to scrap the existing vle/forum object. If that is a large problem, perhaps we have to build on vle/forum instead of http://id.tincanapi.com/activitytype/discussion.

Although we have a vle_resource_viewed statement template (module_viewed), it's not specific to forums and all the examples are about viewing course / module materials. I hoped that we might tweak that, in order to satisfy requirements for the other forum statement templates, namely to identify threads, posts and replies.

The example "forum view" above has significant problems, mainly through design changes since it was created. I suspect that the existing data might have to be amended to meet v1.0, in which case changing to consistent use of http://id.tincanapi.com/activitytype/discussion and http://id.tincanapi.com/activitytype/forum-topic might not be too difficult? That would enable us to put in place a forum_viewed statement template referring to http://id.tincanapi.com/activitytype/discussion in the context, in a fashion consistent with the draft forum post and forum reply statement template.

If we don't need to identify the thread or the post, then we could use:

Actor: student
Verb: create
Object: vle/forum

and 

Actor: student
Verb: reply
Object: vle/forum

These would allow us to pull out that a student had participated by starting a thread or by replying in a particular forum, but little more detail than that. You could capture the actual post using Result, but I'm not sure there would be much point, without knowing the thread.

Statement templates on this simplified pattern would be more straightforward than the ones I've drafted in more detail above. However, they don't meet a requirement for identifying threads and comments, which seems like missing a trick. I'm not suggesting that we should identify relations between the posts!

If it's a problem about using http://id.tincanapi.com/activitytype/discussion rather than vle/forum, I have no problem using vle/forum as such.

I get the feeling that this discussion is boiling down to whether or not vle/forum should consistently be the object in all forum statements. If we're saying that vle/forum should *always* be the object, then it's difficult to capture the threads and comments, because these are more easily understood as the objects of forum posts, while the forum itself is related to the posts hierarchically above in the context. As you can see, I would rather adjust to identifying the forum consistently using contextActivities.parent.

**Comment by ds10** [comment]
moved to feature-vle for 1.1 development. See https://github.com/jiscdev/xapi/commit/e2f08dd2c608fe0b30edd1040ee78634ec20e4e2
